wasm2c/wasm-rt-os-win.c

```diff
@@ -0,0 +1,256 @@
+// Based on
+// https://web.archive.org/web/20191012035921/http://nadeausoftware.com/articles/2012/01/c_c_tip_how_use_compiler_predefined_macros_detect_operating_system#BSD
+// Check for windows (non cygwin) environment
+#if defined(_WIN32)
+
+#include "wasm-rt-os.h"
+#include "wasm-rt.h"
+
+#include <errno.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <windows.h>
+
+#ifdef VERBOSE_LOGGING
+#define VERBOSE_LOG(...) \
+  { printf(__VA_ARGS__); }
+#else
+#define VERBOSE_LOG(...)
+#endif
+
+#define DONT_USE_VIRTUAL_ALLOC2
+
+size_t os_getpagesize() {
+  SYSTEM_INFO S;
+  GetNativeSystemInfo(&S);
+  return S.dwPageSize;
+}
+
+static void* win_mmap(void* hint,
+                      size_t size,
+                      int prot,
+                      int flags,
+                      DWORD alloc_flag) {
+  DWORD flProtect = PAGE_NOACCESS;
+  size_t request_size, page_size;
+  void* addr;
+
+  page_size = os_getpagesize();
+  request_size = (size + page_size - 1) & ~(page_size - 1);
+
+  if (request_size < size)
+    /* integer overflow */
+    return NULL;
+
+  if (request_size == 0)
+    request_size = page_size;
+
+  if (prot & MMAP_PROT_EXEC) {
+    if (prot & MMAP_PROT_WRITE)
+      flProtect = PAGE_EXECUTE_READWRITE;
+    else
+      flProtect = PAGE_EXECUTE_READ;
+  } else if (prot & MMAP_PROT_WRITE)
+    flProtect = PAGE_READWRITE;
+  else if (prot & MMAP_PROT_READ)
+    flProtect = PAGE_READONLY;
+
+  addr = VirtualAlloc((LPVOID)hint, request_size, alloc_flag, flProtect);
+  return addr;
+}
+
+void* os_mmap(void* hint, size_t size, int prot, int flags) {
+  DWORD alloc_flag = MEM_RESERVE | MEM_COMMIT;
+  return win_mmap(hint, size, prot, flags, alloc_flag);
+}
+
+#ifndef DONT_USE_VIRTUAL_ALLOC2
+static void* win_mmap_aligned(void* hint,
+                              size_t size,
+                              int prot,
+                              int flags,
+                              size_t pow2alignment) {
```

Comment:
seems weird to call this specifically `pow2alignment`, since all of the functions require power-of-two alignment